frontend/public/components/utils/resource-icon.tsx

```diff
@@ -2,45 +2,63 @@ import * as React from 'react';
 import * as classNames from 'classnames';
 import * as _ from 'lodash-es';
 
-import { K8sResourceKindReference } from '../../module/k8s';
+import {
+  K8sGroupVersionKind,
+  K8sResourceKindReference,
+  referenceForGroupVersionKind,
+} from '../../module/k8s';
 import { modelFor } from '../../module/k8s/k8s-models';
 import { kindToAbbr } from '../../module/k8s/get-resources';
 
 const MEMO = {};
 
-export const ResourceIcon: React.SFC<ResourceIconProps> = ({ className, kind }) => {
-  // if no kind, return null so an empty icon isn't rendered
-  if (!kind) {
-    return null;
-  }
-  const memoKey = className ? `${kind}/${className}` : kind;
-  if (MEMO[memoKey]) {
-    return MEMO[memoKey];
-  }
-  const kindObj = modelFor(kind);
-  const kindStr = _.get(kindObj, 'kind', kind);
-  const backgroundColor = _.get(kindObj, 'color', undefined);
-  const klass = classNames(`co-m-resource-icon co-m-resource-${kindStr.toLowerCase()}`, className);
-  const iconLabel = (kindObj && kindObj.abbr) || kindToAbbr(kindStr);
-
-  const rendered = (
-    <>
-      <span className="sr-only">{kindStr}</span>
-      <span className={klass} title={kindStr} style={{ backgroundColor }}>
-        {iconLabel}
-      </span>
-    </>
-  );
-  if (kindObj) {
-    MEMO[memoKey] = rendered;
-  }
+export const ResourceIcon: React.SFC<ResourceIconProps> = ({
+  className,
+  groupVersionKind,
+  kind,
+}) => {
+  // if no kind or groupVersionKind, return null so an empty icon isn't rendered
+  if (kind || groupVersionKind) {
+    const kindReference =
+      kind ||
+      referenceForGroupVersionKind(groupVersionKind.group)(groupVersionKind.version)(
+        groupVersionKind.kind,
+      );
+    const memoKey = className ? `${kindReference}/${className}` : kindReference;
+    if (MEMO[memoKey]) {
+      return MEMO[memoKey];
+    }
+    const kindObj = modelFor(kindReference);
+    const kindStr = _.get(kindObj, 'kind', kindReference);
+    const backgroundColor = _.get(kindObj, 'color', undefined);
+    const klass = classNames(
+      `co-m-resource-icon co-m-resource-${kindStr.toLowerCase()}`,
+      className,
+    );
+    const iconLabel = (kindObj && kindObj.abbr) || kindToAbbr(kindStr);
 
-  return rendered;
+    const rendered = (
+      <>
+        <span className="sr-only">{kindStr}</span>
+        <span className={klass} title={kindStr} style={{ backgroundColor }}>
+          {iconLabel}
+        </span>
+      </>
+    );
+    if (kindObj) {
+      MEMO[memoKey] = rendered;
+    }
+
+    return rendered;
+  }
+  return null;
```

Comment:
We should be exiting early whenever possible. Makes code easier to read and avoids the unnecessary indent.